File: functions/contract_proxy/contract_proxy/facades/firestore/bingo_card_cell.py
from contract_proxy.facades.firestore import db
from contract_proxy.models.bingo_card import (
    BingoCardCell,
    BingoUserTokenMintDto,
)

COLLECTION_PREFIX = "bingoCard"


def fetch_bingo_cell(
    bingo_card_id: str, bingo_cell_id
) -> BingoUserTokenMintDto:
    """ビンゴカードに紐づくビンゴセルを取得する

    Args:
        bingo_card_id (str): _description_

    Returns:
        Union[list[BingoCardCell], None]: _description_
    """
    bingo_card = db.fetch(collection=COLLECTION_PREFIX, id=bingo_card_id)
    if bingo_card is None:
        print("bingo_card is not found")
        return None

    bingo_cells = [
        BingoCardCell(**cell) for cell in bingo_card.get("bingoCells")
    ]
    bingo_cell = [cell for cell in bingo_cells if cell.id == bingo_cell_id]

    if len(bingo_cell) == 0:
        print("bingo_cell is not found")
        return None
    target_bingo_cell = bingo_cell[0]
    user = db.fetch(collection="users", id=target_bingo_cell.answered_user)

    response = BingoUserTokenMintDto(
        answered_at=target_bingo_cell.answered_at,
        answered_user=target_bingo_cell.answered_user,
        bingo_cell_id=target_bingo_cell.id,
        image_ai_check_score=target_bingo_cell.imageAiCheckScore
        or 50,  # TODO: 画像AIチェックのスコアがない場合は50にする
        wallet_address=user["walletAddress"],
    )

    return response

File: server/facades/repositories/bingoCard/bingoCardCell.ts
import { BingoCard } from "@/server/models/bingoCard/dto";
import { firestore } from "../../firebase";
import { BingoCellDetail } from "@/server/models/bingoCardCell/dto";

/**
 * ビンゴセルの詳細を取得する
 */
export const getBingoCellDetail = async (
  bingoCardId: string,
  bingoCellId: string
): Promise<BingoCellDetail> => {
  try {
    const docRef = await firestore
      .collection("bingoCard")
      .doc(bingoCardId)
      .get();
    const bingoCard = docRef.data() as BingoCard;
    const bingoCell = bingoCard.bingoCells.find(
      (cell) => cell.id === bingoCellId
    );

    const bingoCardDetail = { ...bingoCell } as BingoCellDetail;

    if (bingoCell?.completed) {
      // ユーザの情報を付与する
      const userDocRef = await firestore
        .collection("users")
        .doc(bingoCell?.answered_user!)
        .get();
      const userInfo = userDocRef.data() as UserInfo;
      bingoCardDetail.answeredUserDetail = userInfo;
    }

    return bingoCardDetail;
  } catch (e) {
    console.error("[getBingoCellDetail]", e);
    throw e;
  }
};

/**
 * ビンゴカードのセルを取得する
 */
export const getBingoCell = async (
  bingoCardId: string,
  bingoCellId: string
) => {
  try {
    const docRef = await firestore
      .collection("bingoCard")
      .doc(bingoCardId)
      .get();
    const bingoCard = docRef.data() as BingoCard;
    const bingoCells = bingoCard.bingoCells;
    const bingoCell = bingoCells.find((cell) => cell.id === bingoCellId);
    return bingoCell;
  } catch (e) {
    console.error("[getBingoCell]", e);
  }
};

/**
 * ビンゴカードのセルを更新する
 */
export const updateBingoCell = async (
  bingoCardId: string,
  bingoCellId: string,
  updateBingoCellDto: UpdateBingoCellDto
) => {
  try {
    // bingoCardIdのドキュメントの中のbingoCellIdのドキュメントを更新する
    const docRef = await firestore
      .collection("bingoCard")
      .doc(bingoCardId)
      .get();
    const bingoCard = docRef.data() as BingoCard;

    // 更新対象のセルを探して更新する
    const bingoCells = bingoCard.bingoCells;
    const newBingoCells = bingoCells.map((cell) => {
      if (cell.id === bingoCellId) {
        cell = {
          ...cell,
          ...updateBingoCellDto,
        };
      }
      return cell;
    });

    // 更新
    const updateDocRef = await firestore
      .collection("bingoCard")
      .doc(bingoCardId)
      .update({
        bingoCells: newBingoCells,
      });

    return bingoCellId;
  } catch (e) {
    console.error("[updateBingoCell]", e);
  }
};
